cedar-policy-generators/src/main.rs

```diff
@@ -0,0 +1,94 @@
+use std::{fs::File, io};
+
+use anyhow::{anyhow, Result};
+use arbitrary::Unstructured;
+use cedar_policy_core::entities::{Entities, TCComputation};
+use cedar_policy_generators::{schema::Schema, settings::ABACSettings};
+use cedar_policy_validator::SchemaFragment;
+use clap::{Args, Parser, Subcommand};
+use rand::{thread_rng, Rng};
+
+#[derive(Parser, Debug)]
+struct Cli {
+    /// Random byte length
+    #[arg(short, long, default_value_t = 4096)]
+    byte_length: u16,
+
+    #[command(subcommand)]
+    command: Commands,
+}
+
+#[derive(Subcommand, Debug)]
+pub enum Commands {
+    /// Generate random hierarchy
+    Hierarchy(HierarchyArgs),
+}
+
+#[derive(Args, Debug)]
+pub struct HierarchyArgs {
+    /// Schema file
+    #[arg(short, long = "schema", value_name = "FILE")]
+    pub schema_file: String,
+    /// Maximum depth
+    #[arg(long, default_value_t = 4)]
+    pub max_depth: usize,
+    /// Maximum width
+    #[arg(long, default_value_t = 4)]
+    pub max_width: usize,
+}
+
+impl From<&HierarchyArgs> for ABACSettings {
+    fn from(value: &HierarchyArgs) -> Self {
+        Self {
+            match_types: false,
+            enable_extensions: true,
+            max_depth: value.max_depth,
+            max_width: value.max_width,
+            enable_additional_attributes: false,
+            enable_like: false,
+            enable_action_groups_and_attrs: true,
+            enable_arbitrary_func_call: false,
+            enable_unknowns: false,
+        }
+    }
+}
+
+fn generate_hierarchy_from_schema(byte_length: usize, args: &HierarchyArgs) -> Result<Entities> {
+    let f = File::open(args.schema_file.clone())?;
+    let fragment = SchemaFragment::from_file(f)?;
+    let mut rng = thread_rng();
+    let mut bytes = Vec::new();
+    bytes.resize_with(byte_length, || rng.gen());
+    let mut u = Unstructured::new(&bytes);
+    let schema = Schema::from_schemafrag(fragment.clone(), args.into(), &mut u)
+        .map_err(|err| anyhow!(format!("failed to construct `Schema`: {:#?}", err)))?;
+    let h = schema
+        .arbitrary_hierarchy(&mut u)
+        .map_err(|err| anyhow!(format!("failed to generate hierarchy: {:#?}", err)))?;
+    Entities::from_entities(
+        h.entities().into_iter().map(|e| e.clone()),
+        TCComputation::ComputeNow,
+    )?;
```

Comment:
This statement is not redundant, which ensures the generated hierarchy doesn't contain cycles.

Reply:
Isn't that done with the `EnforceAlreadyComputed` line?

Reply:
Not really. We don't want to compute TC because it will add TC edges to the hierarchy. Instead, we just want "direct" edges, which mimic entities provided by the users.

Reply:
I'm still not understanding what calling `ComputeNow` and then throwing away the resulting `Entities` object accomplishes?  But I haven't looked into these APIs recently

Reply:
Maybe it's better to use an example. Let's say the generator generates a hierarchy like `A` -> `B`, `B` -> `C`. The ancestors of `A` is just `B` and the ancestors of `B` is `C`. If we run TC computation, we get entities where the ancestors of `A` is `B` and `C`; the ancestors of `B` is `C`. Note the additional edge between `A` and `C`. So, if we want get the raw entities generated without TC edges, we need to throw away the results of TC computation.

Reply:
Yes. Since we need to throw it away, we even call `ComputeNow` here in the first place?

Reply:
> Yes. Since we need to throw it away, why even call `ComputeNow` here in the first place?

I think I want to make sure there are no cycles. I couldn't find the API to detect cycles without computing TC.

Reply:
Does `EnforceAlreadyComputed` not detect cycles?  I think it probably should; if it doesn't, we should file an issue to change that

Reply:
I think it enforces TC are already computed for a hierarchy. So it probably always fails for our raw hierarchy, which doesn't contain TC edges.

Reply:
`EnforceAlreadyComputed` _should_ do what Craig thinks it does. If it doesn't that's a bug. (It's not an exposed API)

Reply:
Yes, it does make sure there are no cycles. But it also enforces TC, which is not we wanted for this function.

Reply:
I see.  I was mistaken, I thought the subsequent call was using `Enforce` but it is actually using `Assume`.  Makes sense.  It would be nice to have a code comment explaining that the call to `Compute` is just to detect cycles, but I'll approve this PR as-is.